tools/autograd/gen_variable_type.py

```diff
@@ -182,52 +181,62 @@ def is_differentiable(arg):
         differentiable_outputs = differentiable_outputs[:1]
 
     requires_derivative = (
-        name not in DONT_REQUIRE_DERIVATIVE and
+        base_name not in DONT_REQUIRE_DERIVATIVE and
         len(differentiable_inputs) > 0 and len(differentiable_outputs) > 0 and
         strategy == 'use_derived')
 
     if func is not None and not requires_derivative:
         print('WARNING: derivative ignored for {}'.format(name), file=sys.stderr)
 
     def setup_derivative():
+        def error_msg():
+            name = declaration['api_name']
+            return '"the derivative for {} is not implemented"'.format(name)
+
+        args_with_derivatives = find_args_with_derivatives()
+
         env = {}
-        body = []
+        env['args_with_derivatives'] = reference_args(args_with_derivatives)
+        env['op'] = func['op'] if func is not None else 'Error'
+        env['op_ctor'] = '' if func is not None else error_msg()
 
-        tensor_args = [arg for arg in arguments if arg['simple_type'] in {'Tensor', 'TensorList'}]
-        tensor_arg_names = [arg['name'] for arg in tensor_args]
-        args_with_derivatives = (
-            tensor_arg_names if func is None else
-            find_args_with_derivatives(tensor_arg_names))
-        body.extend(emit_check_no_requires_grad(tensor_args, args_with_derivatives))
+        setup = []
+        setup.extend(emit_check_output_args())
+        setup.extend(ASSIGN_GRAD_FN.substitute(env).split('\n'))
         if func is not None:
-            env['op'] = func['op']
-            env['op_ctor'] = ''
-            env['save_inputs'] = save_variables(func['saved_inputs'], False)
-            env['args_with_derivatives'] = args_with_derivatives
-        else:
-            env['op'] = 'Error'
-            env['op_ctor'] = '"the derivative for {} is not implemented"'.format(declaration['api_name'])
-            env['save_inputs'] = []
-            env['args_with_derivatives'] = tensor_arg_names
-        body.append(SETUP_DERIVATIVE.substitute(env))
+            setup.extend(save_variables(func['saved_inputs'], False))
+
+        body = []
+        body.extend(emit_check_no_requires_grad(differentiable_inputs, args_with_derivatives))
+        body.append(SETUP_DERIVATIVE.substitute(env, setup=setup))
         return body
 
-    def find_args_with_derivatives(tensor_arg_names):
+    def emit_check_output_args():
+        if not is_out_fn:
+            return []
+        name = declaration['name']
+        args = [arg['name'] for arg in differentiable_outputs]
+        args = '{ ' + ', '.join(args) + ' }'
+        return ['check_output_args("{}", {});'.format(name, args)]
+
+    def find_args_with_derivatives():
         """Find arguments that have derivative definitions"""
+        if func is None:
```

Comment:
when does this come up?

Reply:
If there's no derivative specified. In that case we error in the backwards if any of the inputs (in the forward pass) had requires_grad=True.